proc/proc.go

```diff
@@ -0,0 +1,249 @@
+// proc offers control over arbitrary processes.
+// proc depends heavily on os/exec.
+// If/when proc needs to do resource control & isolation
+// its probable that os/exec will need to be replaced with
+// either the os package itself or syscall.
+package proc
+
+// Fingers crossed that its easier than rewriting os.ForkExec
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"sync"
+	"time"
+)
+
+const outputReadRefreshInterval = time.Millisecond * 10
+
+// Proc is a BitBox process.
+// Stderr and Stdout are dumped to temp files on disk.
+type Proc struct {
+	// TODO: there is 99% chance we need a synchronization primitive at this level
+	stdoutFileName string
+	stderrFileName string
+	cmd            *exec.Cmd
+}
+
+// NewProc constructs and begins process.
+// Stdout & Stderr of the new process are pointed at temp files.
+// The tempfileNames are acessable through the coresponding members.
+func NewProc(cmdName string, args ...string) (Proc, error) {
+
+	var err error
+	var cmdPath string
+	if cmdPath, err = exec.LookPath(cmdName); err != nil {
+		return Proc{}, err
+	}
+	cmd := exec.Command(cmdPath, args...)
+
+	var stdout *os.File
+	if stdout, err = ioutil.TempFile("", ""); err != nil {
+		return Proc{}, err
+	}
+	cmd.Stdout = stdout
+	stdoutName := stdout.Name()
+
+	var stderr *os.File
+	if stderr, err = ioutil.TempFile("", ""); err != nil {
+		return Proc{}, err
+	}
+	cmd.Stderr = stderr
+	stderrName := stderr.Name()
+
+	if err = cmd.Start(); err != nil {
+		return Proc{}, err
+	}
+
+	go func() {
+		// Wait on the cmd to make sure resources get released
+		cmd.Wait()
+	}()
+
+	return Proc{
+		stdoutFileName: stdoutName,
+		stderrFileName: stderrName,
+		cmd:            cmd,
+	}, nil
+}
+
+// Kill causes the running process to exit and closes the related resources.
+func (p Proc) Kill() error {
+	if err := p.cmd.Process.Kill(); err != nil {
+		// processes which have ended return errors
+		return err
+	}
+	// Throw away the error from Wait() because an error is returned if either:
+	// 1. Wait() was already called (cool, we just use it to know that the process exited)
+	// 2. The process returned a non-zero exit code (cool, we will return any exit code)
+	p.cmd.Wait()
+	return nil
+}
+
+// Status returns the status of the process.
+func (p Proc) Status() ProcStatus {
+	if p.cmd.ProcessState == nil {
+		return Running
+	}
+
+	if p.cmd.ProcessState.ExitCode() != 0 {
+		return Stopped
+	}
+	return Exited
```

Comment:
Based on your comments below, this logic doesn't look correct to me.  The docs say:

>  ExitCode returns the exit code of the exited process, or -1 if the process hasn't exited or was terminated by a signal. 

https://golang.org/pkg/os/#ProcessState.ExitCode

Whereas you say:

> // Stopped indicates that the process returned no exit code.

Reply:
Great catch. It took me a minute to realize where my brain was hung up here. I changed the definitions of Stopped and Exited and the check to be in line with `ExitCode()` because that made the most sense to me. How does it look to you now?

Reply:
LGTM.